### PR TITLE
CurveBS: the volume creation should fail when the CurveBS cluster cap…

### PR DESCRIPTION
**Pull request by yubindy**
…acity reaches a certain watermark

Signed-off-by: Zhao ZeYu <2302370185@qq.com>

### What problem does this PR solve?

Issue Number: #1059  

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?):

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

File: src/mds/nameserver2/curvefs.cpp [hunk 1]
```diff
@@ -280,6 +282,49 @@ StatusCode CurveFS::CreateFile(const std::string & fileName,
         }
     }
 
+    std::vector<PoolIdType> logicalPools =
+                topology_->GetLogicalPoolInCluster();
+    logicalPools = topology_->GetLogicalPoolInCluster();
```
Comment (ilixiaocui) [inline src/mds/nameserver2/curvefs.cpp:287]: Why is GetLogicalPoolInCluster() called twice？

File: src/mds/nameserver2/curvefs.cpp [hunk 2]
```diff
@@ -280,6 +282,49 @@ StatusCode CurveFS::CreateFile(const std::string & fileName,
...
+    if (0 == logicalPools.size() && length != 0) {
```
Comment (ilixiaocui) [inline src/mds/nameserver2/curvefs.cpp:288]: The case of length = 0 is not allowed, the judgment here is redundant.

File: src/mds/nameserver2/curvefs.cpp [hunk 3]
```diff
@@ -280,6 +282,49 @@ StatusCode CurveFS::CreateFile(const std::string & fileName,
...
+        LOG(ERROR) << "[ChooseSingleLogicalPool]:"
+                << " Does not have any available logicalPools.";
+        return StatusCode::kFileLengthNotSupported;
+    }
+
+    bool EnoughFlag = false;
+    for (auto pid : logicalPools) {
+    LogicalPool lPool;
```
Comment (ilixiaocui) [inline src/mds/nameserver2/curvefs.cpp:296]: Note the format.

File: src/mds/nameserver2/curvefs.cpp [hunk 4]
```diff
@@ -280,6 +282,49 @@ StatusCode CurveFS::CreateFile(const std::string & fileName,
...
+    if (!topology_->GetLogicalPool(pid, &lPool)) {
```
Comment (ilixiaocui) [inline src/mds/nameserver2/curvefs.cpp:297]: The following code from lines 296 to 312 has the same logic as TopologyChunkAllocatorImpl::ChooseSingleLogicalPool. You can consider putting this part of the logic in topology_, such as encapsulating a GetRemainingSpaceInLogicalPool()

**Comment by yuxubinchen** [comment]
recheck

File: src/mds/nameserver2/chunk_allocator.h [hunk 5]
```diff
@@ -42,6 +43,10 @@ class ChunkSegmentAllocator {
     virtual bool AllocateChunkSegment(FileType type,
         SegmentSizeType segmentSize, ChunkSizeType chunkSize,
         offset_t offset, PageFileSegment *segment) = 0;
+    virtual bool GetRemainingSpaceInLogicalPool(
+        std::vector<PoolIdType>* logicalPools,
+        std::map<PoolIdType, double>* enoughSpacePools,
+        bool chooseone) = 0;
```
Comment (ilixiaocui) [inline src/mds/nameserver2/chunk_allocator.h:49]: std::vector<PoolIdType>* logicalPools  is the input parameter，It is recommended to use const std::vector<PoolIdType> &logicalPools.

**Comment by yuxubinchen** [comment]
recheck

**Comment by yuxubinchen** [comment]
recheck

File: src/mds/nameserver2/chunk_allocator.h [hunk 6]
```diff
@@ -42,6 +43,9 @@ class ChunkSegmentAllocator {
     virtual bool AllocateChunkSegment(FileType type,
         SegmentSizeType segmentSize, ChunkSizeType chunkSize,
         offset_t offset, PageFileSegment *segment) = 0;
+    virtual void GetRemainingSpaceInLogicalPool(
+        const std::vector<PoolIdType>* logicalPools,
```
Comment (ilixiaocui) [inline src/mds/nameserver2/chunk_allocator.h:47]: const std::vector<PoolIdType> &logicalPools,

File: src/mds/nameserver2/chunk_allocator.h [hunk 7]
```diff
@@ -65,6 +69,13 @@ class ChunkSegmentAllocatorImpl: public ChunkSegmentAllocator {
         SegmentSizeType segmentSize, ChunkSizeType chunkSize,
         offset_t offset, PageFileSegment *segment) override;
 
+    void GetRemainingSpaceInLogicalPool(
+        const std::vector<PoolIdType>* logicalPools,
```
Comment (ilixiaocui) [inline src/mds/nameserver2/chunk_allocator.h:73]: ditto

File: src/mds/nameserver2/curvefs.cpp [hunk 8]
```diff
@@ -280,6 +282,25 @@ StatusCode CurveFS::CreateFile(const std::string & fileName,
         }
     }
 
+    std::vector<PoolIdType> logicalPools =
+                topology_->GetLogicalPoolInCluster();
+    std::map<PoolIdType, double> enoughSpace;
+    bool hasEnough = false;
+    chunkSegAllocator_->GetRemainingSpaceInLogicalPool(&logicalPools,
```
Comment (ilixiaocui) [inline src/mds/nameserver2/curvefs.cpp:289]: The return value of `chunkSegAllocator_->GetRemainingSpaceInLogicalPool` is not used. So this function `GetRemainingSpaceInLogicalPool` can directly return void

File: src/mds/nameserver2/curvefs.cpp [hunk 9]
```diff
@@ -280,6 +282,25 @@ StatusCode CurveFS::CreateFile(const std::string & fileName,
         }
     }
 
+    std::vector<PoolIdType> logicalPools =
+                topology_->GetLogicalPoolInCluster();
+    std::map<PoolIdType, double> enoughSpace;
```
Comment (ilixiaocui) [inline src/mds/nameserver2/curvefs.cpp:287]: `remianingSpace` is more suitable than `enoughSpace`.

File: src/mds/nameserver2/curvefs.cpp [hunk 10]
```diff
@@ -280,6 +282,25 @@ StatusCode CurveFS::CreateFile(const std::string & fileName,
...
+    bool hasEnough = false;
+    chunkSegAllocator_->GetRemainingSpaceInLogicalPool(&logicalPools,
+                                &enoughSpace);
+
+    for (auto it = enoughSpace.begin(); it != enoughSpace.end(); it++) {
```
Comment (ilixiaocui) [inline src/mds/nameserver2/curvefs.cpp:292]: A file can be created as long as the accumulated value of all logical pools can exceed length.

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
stop

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
PTAL @ilixiaocui

File: src/mds/nameserver2/curvefs.cpp [hunk 11]
```diff
@@ -280,6 +282,22 @@ StatusCode CurveFS::CreateFile(const std::string & fileName,
         }
     }
 
+    std::vector<PoolIdType> logicalPools =
+                topology_->GetLogicalPoolInCluster();
+    std::map<PoolIdType, double> remianingSpace;
+    uint64_t allRemianingSpace;
```
Comment (ilixiaocui) [inline src/mds/nameserver2/curvefs.cpp:288]: `allRemianingSpace` needs to be initialized.

**Comment by yubindy** [comment]
recheck

**Comment by yubindy** [comment]
recheck

**Review by opencurveadmin — APPROVED** [review]

**Review by ilixiaocui — APPROVED** [review]

**Pull request merged**